rust/agents/relayer/src/msg/metadata_builder.rs

```diff
@@ -0,0 +1,50 @@
+use std::sync::Arc;
+
+use hyperlane_base::{CachingMailbox, ChainSetup, CoreMetrics};
+use hyperlane_core::{
+    accumulator::merkle::Proof, HyperlaneMessage, MultisigSignedCheckpoint, Signers,
+};
+use hyperlane_core::{Mailbox, MultisigIsm, H256};
+
+#[derive(Debug, Clone)]
+pub struct MetadataBuilder {
+    metrics: Arc<CoreMetrics>,
+    signer: Option<Signers>,
+    chain_setup: ChainSetup,
+}
+
+impl MetadataBuilder {
+    pub fn new(
+        metrics: Arc<CoreMetrics>,
+        signer: Option<Signers>,
+        chain_setup: ChainSetup,
+    ) -> Self {
+        MetadataBuilder {
+            metrics,
+            signer,
+            chain_setup,
+        }
+    }
+
+    pub async fn fetch_metadata(
+        &self,
+        message: HyperlaneMessage,
+        mailbox: CachingMailbox,
+        checkpoint: MultisigSignedCheckpoint,
+        proof: Proof,
+    ) -> eyre::Result<Vec<u8>> {
+        let ism_address = mailbox.recipient_ism(message.recipient).await?;
```

Comment:
A lot of this feels cache-able-- doesn't have to be now but this new functionality will increase the # of eth_calls, especially when messages that are failing are retried. I see what you mean now about potentially exposing a verify function which could allow us to cache formatted metadata if a message's ISM.verify call is successful but their handle call isn't.

Maybe not a huge issue now that we have a policy for not retrying old failing messages super frequently